Assets/Scripts/EditorState/JsonSettingState.cs

```diff
@@ -0,0 +1,126 @@
+using System;
+using System.IO;
+using Assets.Scripts.Utility;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Zenject;
+
+namespace Assets.Scripts.EditorState
+{
+    public abstract class JsonSettingState
+    {
+        // Dependencies
+        protected PathState _pathState;
+        protected SceneDirectoryState _sceneDirectoryState;
+
+        [Inject]
+        private void Constructor(PathState pathState, SceneDirectoryState sceneDirectoryState)
+        {
+            _pathState = pathState;
+            _sceneDirectoryState = sceneDirectoryState;
+        }
+
+
+        protected JObject _jsonSettings = null;
+
+        public MyNullable<T> GetSetting<T>(string key)
+        {
+            if (_jsonSettings == null)
+            {
+                LoadSettings();
+            }
+
+            return _jsonSettings!.ContainsKey(key) ?
+                _jsonSettings[key]!.ToObject<T>() :
+                new MyNullable<T>();
+        }
+
+        public void SetSetting<T>(string key, T value)
+        {
+            if (_jsonSettings == null)
+            {
+                LoadSettings();
+            }
+
+            _jsonSettings![key] = JToken.FromObject(value);
+
+            SaveSettings();
+        }
+
+        protected abstract void SaveSettings();
+        protected abstract void LoadSettings();
+    }
+
+    public class ProjectSettingState : JsonSettingState
+    {
+        protected override void SaveSettings()
+        {
+            var projectSettingsDirectoryPath = _pathState.ProjectPath + "/dcl-edit";
```

Comment:
projectSettingsPath is "calculated" here and in the LoadSettings method. Maybe create a single, class-wide method? Something like this:
`string ProjectSettingsPath => _pathState.ProjectPath + "/dcl-edit/settings.json";`